### PR TITLE
Start after nr of consumers for broadcast hub

### PR DESCRIPTION
**Pull request by CloudZou**
Start after nr of consumers for broadcast hub #27163

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @CloudZou,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by raboof** [comment]
OK TO TEST

**Comment by raboof** [comment]
I'm afraid our binary compatibility checker found some incompatibilities:
```
[error]  * method this(Int)Unit in class akka.stream.scaladsl.BroadcastHub does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.BroadcastHub.this")
[error]  * method sink(Int)akka.stream.scaladsl.Sink in object akka.stream.scaladsl.BroadcastHub does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.BroadcastHub.sink")
[error]  * method sink()akka.stream.scaladsl.Sink in object akka.stream.scaladsl.BroadcastHub does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.BroadcastHub.sink")
```

This can probably be fixed by adding deprecated methods with the old parameters.

I also suspect there might be formatting inconsistencies, `sbt scalafmtAll test:scalafmtAll` should fix those

**Comment by CloudZou** [comment]
> I'm afraid our binary compatibility checker found some incompatibilities:
> 
> ```
> [error]  * method this(Int)Unit in class akka.stream.scaladsl.BroadcastHub does not have a correspondent in current version
> [error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.BroadcastHub.this")
> [error]  * method sink(Int)akka.stream.scaladsl.Sink in object akka.stream.scaladsl.BroadcastHub does not have a correspondent in current version
> [error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.BroadcastHub.sink")
> [error]  * method sink()akka.stream.scaladsl.Sink in object akka.stream.scaladsl.BroadcastHub does not have a correspondent in current version
> [error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.scaladsl.BroadcastHub.sink")
> ```
> 
> This can probably be fixed by adding deprecated methods with the old parameters.
> 
> I also suspect there might be formatting inconsistencies, `sbt scalafmtAll test:scalafmtAll` should fix those

thank you very much, this is my first pr for akka, i will fix it.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala [hunk 1]
```diff
@@ -244,50 +259,46 @@ class HubSpec extends StreamSpec {
...
         .concat(Source(2 to 10))
         .throttle(1, 10.millis, 3, ThrottleMode.shaping)
-        .toMat(BroadcastHub.sink(8))(Keep.both)
+        .toMat(BroadcastHub.sink(2, 8))(Keep.both)
         .run()
 
       val f1 = source.runWith(Sink.seq)
       val f2 = source.runWith(Sink.seq)
 
-      // Ensure subscription of Sinks. This is racy but there is no event we can hook into here.
```
Comment (raboof) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala]: :100:

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala [hunk 2]
```diff
@@ -324,7 +335,7 @@ class HubSpec extends StreamSpec {
 
       // sending the first element is in a race with downstream subscribing
       // give a bit of time for the downstream to complete subscriptions
-      Thread.sleep(100)
+//      Thread.sleep(100)
```
Comment (raboof) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala]: Maybe remove these 3 lines completely?

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala [hunk 3]
```diff
@@ -341,17 +352,16 @@ class HubSpec extends StreamSpec {
     }
 
     "properly signal completion to consumers arriving after producer finished" in assertAllStagesStopped {
-      val source = Source.empty[Int].runWith(BroadcastHub.sink(8))
+      val source = Source.empty[Int].runWith(BroadcastHub.sink(1, 8))
       // Wait enough so the Hub gets the completion. This is racy, but this is fine because both
       // cases should work in the end
```
Comment (raboof) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala]: Remove comment as well

File: akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala [hunk 4]
```diff
@@ -443,6 +481,11 @@ private[akka] class BroadcastHub[T](bufferSize: Int)
                   callback.invoke(UnRegister(consumer.id, startFrom, startFrom)))
               case _ => ()
             }
+
+            if (initialized && pending.nonEmpty) {
+              pending.foreach(publish)
```
Comment (raboof) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala]: They were placed in `pending` from `publish`, so it's OK to publish again now that we're initialized.

Might be slightly simpler if we put this into the `if (activeConsumers >= startAfterNrOfConsumers)` block above?

**Review by raboof — APPROVED** [review]
This looks useful! Tests can be cleaned up a little more, but otherwise LGTM

**Comment by hepin1989** [comment]
@CloudZou This need a rebase

**Comment by froth** [comment]
@CloudZou are you still working on this? If not: is it ok if I try to rebase and finish it?
@raboof You approved this a while ago, is this still a valid approach or has the reality changed too much since then?

**Comment by hepin1989** [comment]
> @CloudZou are you still working on this? If not: is it ok if I try to rebase and finish it?
> @raboof You approved this a while ago, is this still a valid approach or has the reality changed too much since then?

I can not reach him on wechat too, I think you can make progress on to of his？or build a new？

**Pull request closed without merging**